models/model_choice.py

```diff
@@ -68,6 +71,55 @@ def load_checkpoint(filename):
         raise FileNotFoundError(f"=> No model found at '{filename}'")
 
 
+def verify_weights(num_classes, weights):
+    """Verifies that the number of weights equals the number of classes if any are given
+    Args:
+        num_classes: number of classes defined in the configuration file
+        weights: weights defined in the configuration file
+    """
+    if num_classes == 1 and len(weights) == 2:
+        warnings.warn("got two class weights for single class defined in configuration file; will assume index 0 = background")
+    elif num_classes != len(weights):
+        raise ValueError('The number of class weights in the configuration file is different than the number of classes')
+
+
+def set_hyperparameters(params, num_classes, model, checkpoint, dontcare_val):
+    """
+    Function to set hyperparameters based on values provided in yaml config file.
+    If none provided, default functions values may be used.
+    :param params: (dict) Parameters found in the yaml config file
+    :param num_classes: (int) number of classes for current task
+    :param model: initialized model
+    :param checkpoint: (dict) state dict as loaded by model_choice.py
+    :return: model, criterion, optimizer, lr_scheduler, num_gpus
+    """
+    # set mandatory hyperparameters values with those in config file if they exist
+    lr = get_key_def('learning_rate', params['training'], None, "missing mandatory learning rate parameter")
+    weight_decay = get_key_def('weight_decay', params['training'], None, "missing mandatory weight decay parameter")
+    step_size = get_key_def('step_size', params['training'], None, "missing mandatory step size parameter")
+    gamma = get_key_def('gamma', params['training'], None, "missing mandatory gamma parameter")
+
+    # optional hyperparameters. Set to None if not in config file
+    class_weights = torch.tensor(params['training']['class_weights']) if params['training']['class_weights'] else None
+    if params['training']['class_weights']:
```

Comment:
I'd suggest, once again, using get_key_def() since it's meant to be a generic function for reading parameters from yaml. Same applies for yaml parameters stored in variables below (loss, optimizer, num_gpus, etc.).

One neat feature you could use here, for example, is the [`expected_type` argument](https://github.com/NRCan/geo-deep-learning/blob/6852729d0ec03c0c2a6465b7a68039181c8bcbda/utils/utils.py#L155) of `get_key_def()`, i.e. expect a `typing.List` (right?) here. It would also help with the exception handling since `verify_weights()` doesn't validate the kind of object being passed in the `weights` argument.

I just saw that I didn't document the expected type argument it in the docstrings. Bad karma.